File: Leetcode/C++/33. Search in Rotated Sorted Array/solution.cpp
// https://leetcode.com/problems/search-in-rotated-sorted-array/

#include <vector>

using namespace std;

class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        int left = 0, right = nums.size() - 1;
        int mid;
        // Find index of most smallest element
        while (left < right)
        {
            /*
            Prevent overflow, left / 2 = left - (left / 2)
            mid = (left + right) / 2 -> left + (right - left) / 2
            */
            mid = left + (right - left) / 2;
            if (nums[mid] > nums[right])
            {
                left = mid + 1;
            }
            else
            {
                right = mid;
            }
        }

        int minIndex = left;
        left = 0;
        right = nums.size() - 1;
        while (left <= right)
        {
            mid = left + (right - left) / 2;
            int midPivot = ((mid) + minIndex) % nums.size();
            if (nums[midPivot] < target)
            {
                left = mid + 1;
            }
            else if (nums[midPivot] > target)
            {
                right = mid - 1;
            }
            else
            { // Found answer
                return midPivot;
            }
        }
        return -1; // Target not in nums
    }
};

File: Leetcode/Python/33. Search in Rotated Sorted Array/solution.py
# https://leetcode.com/problems/search-in-rotated-sorted-array/?envType=study-plan&id=algorithm-ii

# Ref: https://leetcode.com/problems/search-in-rotated-sorted-array/discuss/14437/Python-binary-search-solution-O(logn)-48ms


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        length = len(nums)
        start, end = 0, length - 1

        while start <= end:
            mid = (start + end) >> 1

            if nums[mid] == target:
                return mid
            # left rotated  ex. 3 4 5 1 2
            #                      mid
            if nums[mid] > nums[start]:
                if nums[start] <= target and nums[mid] > target:
                    end = mid - 1
                else:
                    start = mid + 1
            # right rotated  ex. 5 1 2 3 4
            #                       mid
            else:
                if nums[end] >= target and target > nums[mid]:
                    start = mid + 1
                else:
                    end = mid - 1
        return -1  # If target not exist return -1
